README
	Small Potato - xboard/winboard compatible chess engine

  Copyright (C) 2000 Alejandro Dubrovsky

  Small Potato is free software; you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation; version 2 of the License.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program; if not, write to the Free Software
  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

Small Potato used to be called unches.  It still plays under that name in FICS.

This is my attempt at a chess engine.  The code is messy and sparsely 
commented.  It is also quite weak at the moment.  This could all change.

To run using xboard as a frontend, use 'xboard -fcp "smallpotato <options>"'
To run it through WinBoard, you will need to add something like the following to
your winboard.ini:
/fcp "smallpotato <theoptions>" /fd "full\path\to\smallpotato\directory"
/scp "smallpotato <theoptions>" /sd "full\path\to\smallpotato\directory"
If you are using some other interface, then you should read its documentation.

The easier method of configuration now is using the configuration file.  "sp.rc" is
first looked for, and if that doesn't exist, "sp.ini" is opened.  You can also
specify a configuration file through the -config commandline switch.  A default
sp.rc is supplied with instructions on what each switch means.

It is also almost fully configurable through the command line.  Run "smallpotato -help" 
to list them all, or read doc/help.txt if you don't see anything when you run it (some
windows versions).  

Most of the control, though, is through commands during the running  of the program 
(eg. To create or modify a book).  Read commands.txt in the doc directory for more 
information on these.

Small Potato uses clig (version 1.9.9 or greater) to take care of the command
line handling.  It can be gotten from: http://wsd.iitb.fhg.de/~kir/clighome/ clig
is no longer needed for compiling the distribution since cmdline.c and cmdline.h
are being included, but is still needed if you are planning to change the command
line options.

Ideas were stolen/borrowed from lots of places.  Some of them follow:

Crafty
F.D. Laram<E9>e, and his how-to-write-a-chess-program tutorial from gamedev.net
(http://www.gamedev.net/reference/articles/article1014.asp)
GNU Chess
Alpha-Beta & TT by Aske Plaat, Jonathan Schaeffer, Wim Pijls and Arie de Bruin
(http://www.cs.vu.nl/~aske/Papers/tr9417.pdf)
MTD(f) - A Minimax Algorithm faster than NegaScout, by Aske Plaat
Research Re: search & Re-search by Aske Plaat (PhD Thesis) 
(http://www.cs.vu.nl/~aske/Papers/abstr-ks.html)
Computer Chess Web Site by Bruise Moreland
http://www.seanet.com/~brucemo/chess.htm
Pepito <http://www.winboardengines.de/pepito/>
Amy <http://www.stethojupi.de/software.html>
KnightCap <http://samba.org/KnightCap/> and 
http://syseng.anu.edu.au/~jon/papers/knightcap.ps.gz
The Design and Implementation of the Rookie 2.0 Chess Playing Program by M.N.J. van Kervinck
at http://brick.bitpit.net/~marcelk/2002/marcelk-thesis.ps.gz
Rebel's secrets by Ed Schroeder (an umlab on the o) at 
http://members.home.nl/matador/chess840.htm


Comments, code, bugreports and patches are welcome.

	- Alejandro Dubrovsky <alito@organicrobot.com>
	- Small Potato homepage <http://smallpotato.sourceforge.net>


NEWS
25/09/03 - 0.6.1
			Tweaked move ordering (including addition of killer move heuristic)
			Minor tweaking of evaluation (with unknown consequences)
			More configurable settings
			Removed MTD(f) and standard alpha-beta search as search options
			Bugfixes (including K v K draw claim)

17/09/02 - 0.6.0
			Added null-move pruning and aspiration window.
			Improved endgame evaluation (or, more precisely, it now has a separate
			endgame evaluation).  Same with king execution phase.
			Now claims draw on insufficient material.
			Slightly friendlier command line mode.
			Minor bugfixes (fen/epd parsing).

18/08/02 - 0.5.1
			Now can be configured through file.
			It can now resign.
			Switched, and very slightly tweaked, default evaluation function.
			Speeded up by about 20%.

03/08/02 - 0.5.0
			Switched default search algorithm to negascout
			Now recover a PV line for display
			Minor transposition table tweaking
			Minor fixes of bugs introduced in 0.4.0
			Now compiles out-of-the-box in clig-less systems.

17/07/02 - 0.4.0
			Implemented move now and analyze.
			Switched to 0x88 move generation.

11/04/02 - 0.3.3
			Added opening book.
			EPD suites and lines can be solved
			Made it portable to cygwin
			Changed default hashtable size to 32MB.
			Can now specify hashtable size in kilobytes.
			Made aleweights.out be loaded by default

03/04/02 - 0.3.2
			No changes (reflex action release that shouldn't have 
			happened due to unkown last release)

08/03/02 - 0.3.1
			Changed to delayed checking of check for speculated performance
			improvement (even though it does not seem to have materialised)
			Got rid of a couple of often used functions with array lookups for definite
			performance gains.
			Completed switch to move/unmove.

26/02/02 - 0.3.0
			Changed name to Small Potato
			Changed to move/unmove instead of move/copy struct (faster)
			Implemented undo and remove for xboard/protocol 2 programs
			(Note:  Search instability noted in previous version is "meant" to be there.
			Caused by the nature of transposition tables, not by some bug in implementation)
			No bugfixes

07/11/01 - 0.2.4
			Better randomised play
			Switched to 64-bit random hashes
			Lots of bugfixes, including the crash on start 
			introduced in v0.2.3, PGN output bugfixes, and
			some silliness in QuiescentSearch.
			(NOTE:  Search instability still there)

20/10/01 - 0.2.3
			Added history ordering
			Better time management.
			PGN output for game recording.
			Bugfixes

13/10/01 - 0.2.2
			Changed GPL license
			Added quiescent search
			Bugfixes

09/10/01 - 0.2.1
			Added internal iterative deepening
			Bugfixes

03/10/01 - 0.2.0
			Switched to MTD(f)
			Bugfixes

01/10/01 - 0.1.5
			Bit of documentation.
			Fixed the hashtable (maybe)

01/10/01 - 0.1.4
			Improved packaging.
			Couple of bugfixes.

16/04/01 - 0.1.3
			Switched to Zobrist hashing.
			Switched to negamax
			Added runtime evaluation board loading
			Lots of bugfixes (especially hashtable)

14/03/01 - 0.1.2
			Added hashtable.
			Lots of bugfixes.

14/02/01 - 0.1.1
			Added first move move ordering.
			Improved FICS automation.
			Better logging and recording
			Lots of bugfixes

25/12/00 - 0.1.0
			Initial release


board.h
/*
   Copyright (C) 2001 Alejandro Dubrovsky
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; version 2 of the License

 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU Library General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 */

#ifndef BOARD_H
#define BOARD_H 1

#include "common.h"
#include "piecesstruct.h"
#include "communication.h"
#include "hashstruct.h"

#define MAXPLYPERGAME 500
#define MAXEPDMOVES 5
#define KILLERS 2

#define WKCASTLE 1
#define WQCASTLE 1 << 1
#define BKCASTLE 1 << 2
#define BQCASTLE 1 << 3
#define WKCASTLEAVAIL 1 << 4
#define WQCASTLEAVAIL 1 << 5
#define BKCASTLEAVAIL 1 << 6
#define BQCASTLEAVAIL 1 << 7
#define NULLMOVE 1 << 8
#define PALP 1 << 8
#define NORMALMODE 1
#define BOOKMODE 1 << 1
#define FORCEMODE 1 << 2
#define ANALYSEMODE 1 << 3
#define PONDERMODE 1 << 4
#define NORMAL 1
#define ACT 1 << 1
#define MOVE_AND_ACT 1 << 2
#define WAIT_FOR_EVENT 1 << 3
#define MOVE_AND_RESIGN 1 << 4
#define OPENING_PHASE 1
#define END_PHASE 2
#define MATING_PHASE 3

/* Squares (stolen and modified from Phalanx XXII  (didn't feel like typing) */

#define A1 0
#define B1 1
#define C1 2
#define D1 3
#define E1 4
#define F1 5
#define G1 6
#define H1 7

#define A2 16
#define B2 17
#define C2 18
#define D2 19
#define E2 20
#define F2 21
#define G2 22
#define H2 23

#define A3 32
#define B3 33
#define C3 34
#define D3 35
#define E3 36
#define F3 37
#define G3 38
#define H3 39

#define A4 48
#define B4 49
#define C4 50
#define D4 51
#define E4 52
#define F4 53
#define G4 54
#define H4 55

#define A5 64
#define B5 65
#define C5 66
#define D5 67
#define E5 68
#define F5 69
#define G5 70
#define H5 71

#define A6 80
#define B6 81
#define C6 82
#define D6 83
#define E6 84
#define F6 85
#define G6 86
#define H6 87

#define A7 96
#define B7 97
#define C7 98
#define D7 99
#define E7 100
#define F7 101
#define G7 102
#define H7 103

#define A8 112
#define B8 113
#define C8 114
#define D8 115
#define E8 116
#define F8 117
#define G8 118
#define H8 119
#define I8 120

#define OUT 0x88
#define NOPALP -100

struct epdMoveStruct {
	int avoid;   /*1 if to be avoided, 0 if to be sought */
	int count;  /* number of epdmoves */
	Move moves[MAXEPDMOVES];	
};

typedef struct epdMoveStruct epdMove;

struct extMoveDescriptionStruct {
	int from, to;
	pieceType frompiece, topiece, fromcolor, tocolor;
	unsigned int flags;
};

typedef struct extMoveDescriptionStruct extMoveDescription;

struct pvStruct {
	int length;
	Move line[MAXPLYPERGAME];
};

typedef struct pvStruct PV;

struct killerStruct {
	Move move;
	int score;
};

typedef struct killerStruct Killer;

struct boardStruct
{
	timeControl timecontrol;
	hashTable *hashtable; /* transposition table */
	FILE *logfile, *gamefile; /* file for log, and for game recording */
	prodcons *input, *output; /* input and output queues */
	int (*evalboard) ();  /* evaluation function to use */
	int (*engine) ();  /* search engine to use */
	int (*searcher) ();  /* searching bit (MTDf, negascout) */
	int betaers[2][128][128]; /* history list of beta-cutting moves */
	int piecevalues[NOTHING - WKING][128]; /* piece-value tables for current phase */
	int piececount[NOTHING - WKING]; /* number of pieces of each type left*/
	PV pv; /* the principal variation */
	char *name;  /* engine name */
	epdMove epdmove;  /* the epdmove/s to look for/avoid */
	extMoveDescription game[MAXPLYPERGAME]; /* history of moves, so that we can unmove */
	pieceType squares[128]; /* piece on each square */
	pieceType colors[128]; /* color of each square */
	Action lastaction;
	Killer killers[MAXPLYPERGAME][KILLERS]; /* killer moves */
	U64 history[MAXPLYPERGAME];  /* hash of board at each ply */
	int fifty[MAXPLYPERGAME];  /* number of plys since last pawn was moved, or piece taken at each ply*/
	int palp[MAXPLYPERGAME];  /* en-passant numbers at each ply */
	int blacktotal, whitetotal; /* total value for black and white (by piecevalues above) */
	int wkcastle, bkcastle, wqcastle, bqcastle; /* rights to each castle */
	int blackturn;  /* 1 for black's turn, 0 for not */
	int ply;  /* ply number of the game */
	int bkingpos, wkingpos;  /* position of white and black king */
	int ourtime;  /* hundredths of seconds on our clock */
	int theirtime; /* hundredths of seconds on their clock */	
	int totalnodes, totalevals, totalhash, totalquiescent; /* general counters */
	int lastcheck;  /* when last time check was made */
	int post;  /* post thinking */
	int analyse;  /* analyse mode */
	int depth;  /* maximum depth to search */
	int verbosity;  /* how much are we throwing to the log */
	int record; /* are we recording the game */
	int resign; /* evaluation at which we should resign */
	int timelimit;  /* maximum number of hundredths of seconds per move */
	int random;	 /* how much noise to add to the value tables before each game */
	int actionmode; /* what are we doing */
	int ponder; /* do we ponder if we can */
	int anomaly; /* as defined above (NORMAL, ACT, MOVE_AND_ACT) */
	int phase; /* what phase of the game we are in (eg opening, middle, endgame) */
	int nulldepth; /* denotes last depth at which nullmove was taken */
	int epdmindepth;  /* don't return from an epd until at least this depth has been reached */	
	int drawvalue; /* value of a draw */
	int timeaggression; /* a number to denote how much time to take on each move, which at
	present represents the minimum percentage of time of an "average move" */
};

typedef struct boardStruct Board;

void setColorTotals(Board * board);
void initialiseBoard(Board * board);

void movePieceSimple(Board * board, const Move * const move);
void movePiece(Board * board, const Move * const move);
void unMove(Board * const board);
void unMoveSimple(Board * const board);

void makeNullMove(Board * const board);
void unmakeNullMove(Board * const board);

int leftInCheck(Board *board, Move *move);
int isLegal(Board * board, Move * move);
int inCheck(const Board * const board);
int isCheckmate(Board * board);
int isStalemate(Board * board);
int isRepetitionDraw(Board * board);
int isInsufficientMaterialDraw(Board * board);
int generateMoveList(Board * const b, Move *movelist);
int generateTentativeMoveList(const Board * const b, Move *movelist);
int generateQuiescentMoveList(Board * const b, Move *movelist);
int generateTentativeQuiescentMoveList(const Board * const b, Move *movelist);
int checkRepetitionDraw(Board * board);
int canMove(Board * const board);

/*
inline int x88ToStandard(const int x88);
inline int standardTox88(const int standard);
inline int x88PosToRow(const int pos);
inline int x88PosToCol(const int pos);
inline int posToRow(const int pos);
inline int posToCol(const int pos);
inline int rowColToPos(const int row, const int col);
inline int x88RowColToPos(const int row, const int col);
*/

static inline int rowColToPos(const int row, const int col)
{
	return (row << 3) + col;
}

static inline int x88RowColToPos(const int row, const int col)
{
	return (row << 4) + col;
}


static inline int posToRow(const int pos)
{
	return (pos >> 3);
}

static inline int posToCol(const int pos)
{
	return (pos & 7);
}

static inline int x88PosToRow(const int pos)
{
	return (pos >> 4);
}

static inline int x88PosToCol(const int pos)
{
	return (pos & 7);
}



static inline int x88ToStandard(const int x88)
{
	return (x88 & 7) | ((x88 & 0x70) >> 1);
}

static inline int standardTox88(const int standard)
{
	return (standard & 7) | ((standard & 0x38) << 1);
}


#endif
